EIPS/eip-7748.md

```diff
@@ -0,0 +1,292 @@
+---
+eip: 7748
+title: State conversion to Verkle Tree
+description: Describes a state conversion procedure to migrate key-values from the Merkle Patricia Tree to the Verkle Tree.
+author: Guillaume Ballet (@gballet), Ignacio Hagopian (@jsign), Gajinder Singh (@g11tech), Ansgar Dietrichs (@adietrichs), Gottfried Herold (@GottfriedHerold), Jamie Lokier (@jlokier), Tanishq Jasoria (@tanishqjasoria), Parithosh Jayanthi (@parithosh), Gabriel Rocheleau (@gabrocheleau), Karim Taam (@matkt)
+discussions-to: https://ethereum-magicians.org/t/eip-7748-state-conversion-to-verkle-tree/20625
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-07-23
+requires: 7612
+---
+
+## Abstract
+
+This EIP proposes a procedure to convert, on each block, a fixed number of key-values from the existing  Merkle Patricia Tree (MPT) to the Verkle Tree (VKT).
+
+## Motivation
+
+The accounts state is too large to wait for transactions to organically move all of them to the VKT through the Overlay Tree. Thus, we need a strategy to convert all the state within a reasonable time. The state conversion completion allows removing the Overlay Tree abstraction introduced in [EIP-7612](./eip-7612.md) and use directly the VKT for all state access.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Constants
+
+| Parameter                    | value | Description                                                    |
+| ---------------------------- | ----- | -------------------------------------------------------------- |
+| `CONVERSION_START_TIMESTAMP` | `TBD` | Timestamp at which the conversion starts.                      |
+| `CONVERSION_STRIDE`          | `TBD` | Maximum number of _conversion units_ to be converted per block |
+
+A _conversion unit_ is:
+
+- A contract storage slot.
+- A contract code. (i.e. all the code is a single _conversion unit_)
+- An account data. (e.g. balance, nonce, code-hash)
```

Comment:
So, would "balance" be a conversion unit, or would "balance+nonce+codehash" be a conversion unit? And how would that play out with the updated verkle leaf design where all version, balance, nonce and codesize are all package within a single slot (BASIC_DATA)? It seems to me like it would make sense to consider a "unit" whatever fits in a single "verkle slot" in this case.

Reply:
A conversion unit is "An account data" which includes all the fields (balance, nonce, code-hash). The rationale for this is that on the MPT side all these values are packed in a single tree key.

The complete space of options also have other variants. As you mention can be thought as "what is a unit" on the VKT side. Or maybe think of each field as a unit. And maybe others.

The option of "what is a unit on the MPT side" is the one chosen here. I see those other options as making the logic more complex, since now it means there're more places where you can partially migrate an account. And I don't see the benefit.

Reply:
Yeah that makes sense. Definitely, I think we should go for whatever has the smallest complexity burden, and partial account migration is a concern that I had in mind. If we imagine an edge case where an account gets partially migrated at block `n`:
- balance and nonce get migrated, and then we stop the account conversion short of converting `code_size`, having reached "n = CONVERSION_TRIDE` 

Then we have a situation where the "BASIC_DATA" of the verkle tree has been partially updated. We therefore have to keep track of the fact that some BASIC_DATA fields have been converted (e.g. version, balance, nonce), while code_size has not. Clients will therefore have to make that discernment during the transition, and won't be able to rely on e.g. the simple fact that `BASIC_DATA.version !== undefined`. What do you think?

Reply:
That's exactly the reason why I prefered the current `AccountDataPhase`. All the fields are moved in one go and there's no partial migration for all those fields.

Currently, the `***Phase` describe all the situations where you can have a partially migrated account. Today we only have `StoragePhase` and `AccountDataPhase` so to push as much as possible to reduce these "partially migrated" situations.

So I believe we're both saying the same? I didn't quite get if your last question implies you're proposing something different than today.

Reply:
Ah it sounds like we agree then! Good